tests/epyccel/test_generic_functions.py

```diff
@@ -418,3 +417,32 @@ def test_zeros_types(language):
 
     assert fl_1 == fl_2
     assert isinstance(fl_1, type(fl_2))
+
+@pytest.mark.parametrize( 'language', (
+        pytest.param("fortran", marks = [
+            pytest.mark.xfail(reason="In generic interface 'scalar_or_array' at (1) procedures must be either all SUBROUTINEs or all FUNCTIONs. Using subroutines first requires issue #297"),
+            pytest.mark.python]
```

Comment:
Fixed 5bbd41b4

Reply:
We need some way of indicating that the function is not a `function` but a `subroutine`. When I dug into this issue I thought that the simplest way of handling this would be to change the prototype and indicate this in the header file. However taking another look, I think there are other ways of doing this which could allow us to fix this now. Do you think I should fix this as part of this PR? Or create a new issue describing a different way of fixing it? Or leave it for now?

Reply:
As it is a Fortran only problem I think it could be handled in the Fortran printer by checking if the `FunctionDef` is an attribute of an `Interface`. This would then cause problems for the wrapper. But we could decide that a `BindCFunctionDef` is never a `function` as it often passes additional data anyway. This would be easy to manage in my future PR for issue #929 which will add the `FortranToCWrapper` class

Reply:
If I understand correctly, you are suggesting that a templated Python function should be translated to only subroutines in Fortran, accessible through an `interface`, if any one of the type combinations require it.

I am afraid that this may be inefficient for very small functions which appear in a right-hand-side expression.

How about creating two different interfaces instead? One would be for subroutines and the other for functions.

Reply:
That sounds like a more sensible suggestion. However I think this suggestion would be much much easier after #297 as it would allow us to save the two interfaces separately in the header. Additionally we have previously said that one of the main motivations for #297 is to allow functions to be renamed and have an easy way to collect the original Python name

Reply:
All right, I suggest that we write a dedicated issue then.

Reply:
https://github.com/pyccel/pyccel/issues/1339 06bd47e3